platform/mds/mds/src/main/java/org/motechproject/mds/service/impl/UserPreferencesServiceImpl.java

```diff
@@ -0,0 +1,199 @@
+import java.util.List;
+
+/**
+ * Implementation of the {@link org.motechproject.mds.service.UserPreferencesService}.
+ */
+@Service
+public class UserPreferencesServiceImpl implements UserPreferencesService {
+
+    private AllUserPreferences allUserPreferences;
+
+    private AllEntities allEntities;
+
+    private SettingsService settingsService;
+
+    @Override
+    public List<UserPreferencesDto> getEntityPreferences(Long id) {
+        Entity entity = getEntity(id);
+        List<UserPreferences> userPreferences =  allUserPreferences.retrieveByClassName(entity.getClassName());
+        List<UserPreferencesDto> dtos = new ArrayList<>();
+
+        if (userPreferences != null) {
+            Integer defaultGridSize = settingsService.getGridSize();
+            for (UserPreferences preferences : userPreferences) {
+                UserPreferencesDto dto = preferences.toDto();
+                if (dto.getGridRowsNumber() == null) {
+                    dto.setGridRowsNumber(defaultGridSize);
+                }
+                dtos.add(dto);
+            }
+        }
+
+        return dtos;
+    }
+
+    @Override
+    @Transactional
+    public UserPreferencesDto getUserPreferences(Long id, String username) {
+        Entity entity = getEntity(id);
+        UserPreferences userPreferences =  allUserPreferences.retrieveByClassNameAndUsername(entity.getClassName(), username);
+        userPreferences = checkPreferences(userPreferences, entity, username);
+
+        if (userPreferences != null) {
+            UserPreferencesDto dto = userPreferences.toDto();
+            if (dto.getGridRowsNumber() == null) {
+                dto.setGridRowsNumber(settingsService.getGridSize());
+            }
+            return dto;
+        }
+
+        return null;
+    }
+
+    @Override
+    @Transactional
+    public void updateGridSize(Long id, String username, Integer newSize) {
+        Entity entity = getEntity(id);
+        UserPreferences userPreferences =  allUserPreferences.retrieveByClassNameAndUsername(entity.getClassName(), username);
+        userPreferences = checkPreferences(userPreferences, entity, username);
+
+        if (newSize == null) {
+            userPreferences.setGridRowsNumber(settingsService.getGridSize());
+        } else {
+            userPreferences.setGridRowsNumber(newSize);
+        }
+        allUserPreferences.update(userPreferences);
+    }
+
+    @Override
+    @Transactional
+    public void selectField(Long id, String username, String fieldName) {
+        Entity entity = getEntity(id);
+        UserPreferences userPreferences =  allUserPreferences.retrieveByClassNameAndUsername(entity.getClassName(), username);
+        userPreferences = checkPreferences(userPreferences, entity, username);
+
+        Field field = entity.getField(fieldName);
+        assertField(field);
+        List<Field> fields = userPreferences.getVisibleFields();
+        fields.add(field);
+        userPreferences.setVisibleFields(fields);
```

Comment:
set is probably not required since you are manipulating the orignal collection (maybe it would be better to just add an addField method to UserPreferences though)